Add logic to resolve implicit versions of Microsoft.AspNetCore.App

JunTaoLuo opened the pull request:
Addresses https://github.com/aspnet/Universe/issues/967.

We are adding logic to resolve implicit version numbers for the Microsoft.AspNetCore.App metapackage in aspnet apps which uses the Web SDK. This is for rc1 and I will send a shiproom email for ask mode. Please continue reviewing the PR but I will not merged until approved.

I'm open to implementation and naming suggestions but I've been trying to keep in line with what I have seen in the dotnet SDK, https://github.com/dotnet/sdk/pull/2085/files.

JunTaoLuo commented on build/package.proj:
```diff
@@ -44,6 +47,38 @@
+  <Target Name="_GenerateBundledVersionsProps">
+    <PropertyGroup>
+      <BundledVersionsPropsFileName>BundledVersions.props</BundledVersionsPropsFileName>
+      <BundledVersionsPropsDirectory>$(WebSdkRoot)/src/Web/Microsoft.NET.Sdk.Web.Targets/</BundledVersionsPropsDirectory>
```
This is the simplest way to generate and pack this file though I don't like the fact that this will generate a file in src/ which is tracked by version control by default. I'm not sure I want to setup the infrastructure to pack in a different working directory to add more complexity to this PR but maybe I should add the specific file to .gitignore?

JunTaoLuo commented on src/Web/Microsoft.NET.Sdk.Web.Targets/Sdk.DefaultItems.targets:
```diff
@@ -0,0 +1,39 @@
+		Condition="Exists('$(MSBuildThisFileDirectory)BundledVersions.props')" />
+
+  <Choose>
+    <When Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' AND '$(_TargetFrameworkVersionWithoutV)' == '2.1'" >
```
I'm doing a minimal implementation here for now. Once we start to patch 2.1 (which will require separate versions for self contained and portable apps) and when we branch for 2.2 (multiple versions depending on tfm), we will need to update the implementation here. Should I implement that right now or should I wait to add it until needed?

natemcmaster replied:
Better to add it now to make sure it works IMO. Functional changes in patches are harder to get approved.

JunTaoLuo replied:
There's very little that can be tested since right now the versions for portable and self contained apss are the same, but I'll add the logic anyway.

JunTaoLuo commented:
Tested manually on the command line and VS. Unit testing will be added in the Templating repo.

JunTaoLuo commented on src/Web/Microsoft.NET.Sdk.Web.Targets/Sdk.DefaultItems.targets:
```diff
@@ -0,0 +1,39 @@
+  <Choose>
+    <When Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' AND '$(_TargetFrameworkVersionWithoutV)' == '2.1'" >
+      <PropertyGroup>
+        <AspNetCoreSharedFrameworkPackageName Condition="'$(AspNetCoreSharedFrameworkPackageName)' == ''">Microsoft.AspNetCore.App</AspNetCoreSharedFrameworkPackageName>
```
Hmmm, maybe comments here would help.

vijayrkn commented on build/package.proj:
```diff
@@ -23,7 +26,7 @@
   <Target Name="_Package" DependsOnTargets="$(_PackageDependsOn)" />
 
-  <Target Name="_CreateNuGetPackages">
+  <Target Name="_CreateNuGetPackages" DependsOnTargets="_GenerateBundledVersionsProps">
```
Why not move this to the above PropertyGroup so that all the dependencies are in one place and easy to find?

```
     <_PackageDependsOn>
       $(_PackageDependsOn);
         _GenerateBundledVersionsProp;
       _CreateNuGetPackages;
     </_PackageDependsOn>
```

JunTaoLuo replied:
Logically _CreateNuGetPackages requires _GenerateBundledVersionsProps to run otherwise it would be missing files, hence the DependsOnTargets. I'll add _GenerateBundledVersionsProps to _PackageDepends on so it's clear that it's also run during pack.

vijayrkn commented on pack/Microsoft.NET.Sdk.Web/Microsoft.NET.Sdk.Web.csproj:
```diff
@@ -18,11 +18,11 @@
     <PackageReference Include="Microsoft.NET.Sdk.Publish" Version="$(PackageVersion)" />
 
-    <Content Include="$(WebSdkSource)\Web\Microsoft.NET.Sdk.Web.Targets\Sdk.props">
+    <Content Include="$(WebSdkSource)\Web\Microsoft.NET.Sdk.Web.Targets\*.props">
```
Will need changes here also - https://github.com/aspnet/websdk/blob/master/build/test.proj#L42

We copy these targets to the local cli install folder so that we can run the end to end tests against the new bits.

natemcmaster commented on build/package.proj:
```diff
@@ -44,6 +47,38 @@
+  -->
+  <Project>
+    <PropertyGroup>
+      <BundledMicrosoftAspNetCoreAppPackageVersion>$(MicrosoftAspNetCoreAppPackageVersion)</BundledMicrosoftAspNetCoreAppPackageVersion>
```
Should we also add a variable for .All?

natemcmaster commented on src/Web/Microsoft.NET.Sdk.Web.Targets/Sdk.DefaultItems.targets:
```diff
@@ -0,0 +1,39 @@
+    <When Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' AND '$(_TargetFrameworkVersionWithoutV)' == '2.1'" >
+      <PropertyGroup>
+        <AspNetCoreSharedFrameworkPackageName Condition="'$(AspNetCoreSharedFrameworkPackageName)' == ''">Microsoft.AspNetCore.App</AspNetCoreSharedFrameworkPackageName>
+        <AspNetCoreSharedFrameworkVersion Condition="'$(AspNetCoreSharedFrameworkVersion)' == ''">$(BundledMicrosoftAspNetCoreAppPackageVersion)</AspNetCoreSharedFrameworkVersion>
```
For consistency with `RuntimeFrameworkVersion`, which is used for NETCore.App, how about `AspNetCoreRuntimeFrameworkVersion`?

natemcmaster commented on src/Web/Microsoft.NET.Sdk.Web.Targets/Sdk.DefaultItems.targets:
```diff
@@ -0,0 +1,39 @@
+      </PropertyGroup>
+
+      <ItemGroup>
+        <AspNetCoreAppPackageReference Include="@(PackageReference->WithMetadataValue('Identity', '$(AspNetCoreSharedFrameworkPackageName)'))" />
```
Consider naming these item groups with an underscore to indicate they are meant for internal use only.

natemcmaster commented on src/Web/Microsoft.NET.Sdk.Web.Targets/Sdk.targets:
```diff
@@ -6,17 +6,20 @@ WARNING:  DO NOT MODIFY this file unless you are knowledgeable about MSBuild and
   <Import Sdk="Microsoft.NET.Sdk.Publish" Project="Sdk.targets" />
 
+  <Import Project="$(MSBuildThisFileDirectory)Sdk.DefaultItems.targets"
+		Condition="Exists('$(MSBuildThisFileDirectory)Sdk.DefaultItems.targets')" />
```
Why the condition? Shouldn't this file always exist as well?

natemcmaster commented on src/Web/Microsoft.NET.Sdk.Web.Targets/Sdk.DefaultItems.targets:
```diff
@@ -0,0 +1,39 @@
+          <Version>$(AspNetCoreSharedFrameworkVersion)</Version>
+          <IsImplicitlyDefined>true</IsImplicitlyDefined>
+        </PackageReference>
+        <ExpectedPlatformPackages Include="$(AspNetCoreSharedFrameworkPackageName)" ExpectedVersion="$(AspNetCoreSharedFrameworkVersion)" />
```
What is this item group for?

dsplaisted replied:
It's to generate an error if you restore for one version of the reference but build with another one.  This would happen if you restore as framework dependent, and try to build for self-contained.  See https://github.com/dotnet/sdk/pull/2148 and https://github.com/dotnet/designs/pull/36.

natemcmaster commented on src/Web/Microsoft.NET.Sdk.Web.Targets/Sdk.DefaultItems.targets:
```diff
@@ -0,0 +1,39 @@
+          Update="$(AspNetCoreSharedFrameworkPackageName)"
+          Condition="'@(AspNetCoreAppPackageReference->Count())' == '1' AND '@(ExplicitAspNetCoreAppPackageReference->Count())' == '0'">
+          <Version>$(AspNetCoreSharedFrameworkVersion)</Version>
+          <IsImplicitlyDefined>true</IsImplicitlyDefined>
```
Should we also be setting PrivateAssets='All'?

dsplaisted replied:
IMO: Yes

dsplaisted commented on build/package.proj:
```diff
@@ -44,6 +47,38 @@
+
+  </Target>
+
+  <Target Name="_GenerateBundledVersionsProps">
```
I do not think you should create a bundled versions file in this repo.  I think you should just hard code a default for the version in Sdk.DefaultItems.targets, if it's not already set.  Then update the bundled versions file in dotnet/cli to include the ASP.NET Core version.

If you put this version in this repo you will have to update the web SDK every time there's a new version of ASP.NET Core.  That will probably be hard to sync when shipping preview and patch versions of ASP.NET Core.

JunTaoLuo replied:
My understanding from the dotnet sdk is to set the versions here for testing and then allow CLI to override. However, now that I think about it, our unit tests will be in a different repo and I don't think there will be anything broken here without default versions. I'll check to make sure but maybe we can remove this component.

JunTaoLuo commented on build/package.proj:
```diff
@@ -1,6 +1,9 @@
 <Project ToolsVersion="14.0" DefaultTargets="Package" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
 
+  <Import Project="$(MSBuildThisFileDirectory)dependencies.props"
+		Condition="Exists('$(MSBuildThisFileDirectory)dependencies.props')" />
```
I'm following the pattern I'm seeing in this repo but I agree these files should always exist.

mmitche commented on build/dependencies.props:
```diff
@@ -0,0 +1,6 @@
+  <!-- These package version may be overridden or updated by automation. -->
+  <PropertyGroup Label="Package Versions: Auto" Condition=" '$(DotNetPackageVersionPropsPath)' == '' ">
+    <MicrosoftAspNetCoreAppPackageVersion>2.1.0-preview2-30563</MicrosoftAspNetCoreAppPackageVersion>
+  </PropertyGroup>
```
@natemcmaster Is the package version props download logic in place?

natemcmaster replied:
Good point. @JunTaoLuo you will probably need to add a parameter, of one does not yet exist, that will download PB_PackageVersionPropsUrl to a local file path. For example, in aspnet/Universe: 

https://github.com/aspnet/Universe/blob/827958e7f244ffdfdbbe219ea1b966830885c413/run.ps1#L208-L214

mmitche commented:
This should be tested locally using a real package version props and restore source URL before check-in, to ensure everything was done properly for dependency uptake.

JunTaoLuo commented:
Updated. Tested locally using command line, VS and VS Code for restore, run, publish and standalone publish. Updated corresponding tests in https://github.com/aspnet/templating/pull/479.

vijayrkn commented:
@JunTaoLuo - Can you please re-target this PR against release/2.1 branch?

JunTaoLuo commented:
Ah so this follows the aspnet branch scheme instead of the cli branch scheme?

JunTaoLuo commented:
ping @natemcmaster @vijayrkn @dsplaisted

dsplaisted commented on src/Web/Microsoft.NET.Sdk.Web.Targets/Sdk.DefaultItems.targets:
```diff
@@ -0,0 +1,143 @@
+    <When Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'" >
+
+      <!-- Default patch version of .All Framework -->
+      <PropertyGroup Condition="'$(DefaultAspNetCoreAllPatchVersion)' == ''">
```
If you are going to support the .All package as well as the .App package, then I'd suggest also supporting it when targeting .NET Core / ASP.NET Core 2.0.  Then you'll be able to to test the roll-forward logic when the target framework doesn't match the bundled version, making it more likely that everything will work when we start working on the next minor version of ASP.NET Core.

JunTaoLuo replied:
The mechanism for the .All package is different for 2.0 since it uses the runtime store and not the shared framework. It would be a major design change and isn't something we want to do right now.
